x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtAuthenticationToken.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+package org.elasticsearch.xpack.security.authc.jwt;
+
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jose.util.Base64URL;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+
+import org.elasticsearch.common.settings.SecureString;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xpack.core.security.authc.AuthenticationToken;
+import org.elasticsearch.xpack.security.authc.BearerToken;
+
+import java.text.ParseException;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * An {@link AuthenticationToken} to hold JWT authentication related content.
+ */
+public class JwtAuthenticationToken extends BearerToken {
+    // Stored members (Note: also includes super.bearerString, which is a SecureString too)
+    protected final SecureString clientAuthorizationSharedSecret; // optional, nullable
+
+    // Processed members (ParseException)
+    protected final AtomicReference<SignedJWT> signedJwt = new AtomicReference<>(null);
+    protected final AtomicReference<JWSHeader> jwsHeader = new AtomicReference<>(null);
+    protected final AtomicReference<JWTClaimsSet> jwtClaimsSet = new AtomicReference<>(null);
+    protected final AtomicReference<Base64URL> jwtSignature = new AtomicReference<>(null);
```

Comment:
JwtAuthenticationToken started out by extending BearerToken and adding a second SecureString. I extended clearCredentials() to call clientAuthorizationSharedSecret.close().

Later, I added parsed JWT content objects. It didn't make sense to call SecureString.close() without also trying to clear the JWT content objects, so I wrapped those too.

I considered options for wrapping like Optional, AtomicReference, etc. Based on some examples I found in the code, I went with AtomicReference.

Reply:
@tvernum is this explanation OK?

Reply:
Does this preclude them being regular fields? They wouldn't be `final`, but they're effectively not final now since the reference can be changed.

That is, it seems to more obvious to have:
```
     protected Base64URL jwtSignature = null; 
```

Reply:
Resolved. I removed usage of AtomicReference wrapping an final qualifiers.